### PR TITLE
Fix check behavior on restart

### LINKED ISSUE #4364: [dev.icinga.com #12148] Checks should not result in "terminated" when restarting an Icinga node
This issue has been migrated from Redmine: https://dev.icinga.com/issues/12148

**Created by tgelf on 2016-07-13 16:38:35 +00:00**

Assignee: _(none)_
Status: _New_ 
Target Version: _Backlog_
Last Update: _2016-07-25 15:37:39 +00:00 (in Redmine)_

    Icinga Version: 2.4.10
    Backport?: Not yet backported
    Include in Changelog: 1

---

Restarting a busy Icinga instance results in lots of checked objects turning there state into unknown, with a plugin output of:

This should neither have an influence on current state nor on state history at all.

### PR DESCRIPTION
**Pull request by Crunsher**
This patch changes the way checkresults are handled during a restart.

  1. Check results coming in during a shutdown are ignored.
  2. Upon start, checks which should have ran (next_check in the past),
  are re-scheduled within the first minute.

This new behavior means there will be no more "Unknown - Terminated"
checkresults during a restart and checks with high check_interval will
be run earlier if they were already scheduled to run. The downside is
that after Icinga2 was down for a while, there will be a lot of checks
within the first minute. Our max concurrent check should take care of
this though.

fixes #4364

### CODE REVIEW

**Comment by Crunsher** [comment]
TODOS:

- [ ] Document this, a note akin to the commit message
- [ ] Further testing, changes to the check logic should be tested especially thorough

**Comment by Crunsher** [comment]
Somehow the tests work on my machine just fine :woman_shrugging: 
Maybe Appveyor is acting up again

File: lib/icinga/checkable-check.cpp [hunk 1]
```diff
@@ -428,8 +428,6 @@ void Checkable::ExecuteCheck()
 	double scheduled_start = GetNextCheck();
 	double before_check = Utility::GetTime();
 
-	UpdateNextCheck();
```
Comment (dnsmichi) [inline lib/icinga/checkable-check.cpp]: This change is problematic, this line ensures that Icinga expects the next check at a specific time. The whole freshness logic is built upon this. Please verify and test that freshness still works with this commit.

Comment (Crunsher) [reply lib/icinga/checkable-check.cpp]: `UpdateNextCheck()` is still called in `ProcessCheckResult()` for active checks, updating the next check in the check execution seems very premature (And would prevent the rest of this patch from working)

**Comment by dnsmichi** [comment]
Discussed and cleared it offline, will further test it.

**Comment by dnsmichi** [comment]
So far, I cannot observe any problems with the patch. Still, this needs deeper tests, and should reach a broader audience from snapshot packages.

**Comment by dnsmichi** [comment]
This likely fixes @lippserd 's #4436 too.

**Pull request merged**